### PR TITLE
Makes Skia's vkQueueSubmit threadsafe.

### PR DESCRIPTION
**Pull request by gaaclarke**
fixes https://github.com/flutter/flutter/issues/133933

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/embedder/tests/embedder_unittests.cc [hunk 1]
```diff
@@ -995,6 +1010,133 @@ TEST_F(EmbedderTest, NoLayerCreatedForTransparentOverlayOnTopOfPlatformLayer) {
   ASSERT_EQ(context.GetSurfacePresentCount(), 0u);
 }
 
+#if SHELL_ENABLE_VULKAN
+TEST_F(EmbedderTest,
```
Comment (gaaclarke) [inline shell/platform/embedder/tests/embedder_unittests.cc:1014]: @bdero please advice how you want these tests to be.  They are just straight copies of the software rendering now.  It was the easiest way I could find to test this.

Comment (bdero) [reply shell/platform/embedder/tests/embedder_unittests.cc:1014]: Oh interesting -- does this test reproduce the problem and fail without your change? And if not, is `CanRenderGradientWithCompositorOnNonRootLayer` sufficient for covering the code in this PR?

If the test doesn't repro the problem, I would just remove it. Otherwise, tossing it into `embedder_gl_unittests` and writing it using `EmbedderTestMultiBackend` to cover both GL and Vulkan seems like a good approach for the moment.
Later on we should add a software context support to the multi-backend class, but that's perhaps a job for another day.

Basically, anything that's Vulkan-specific has a switchboard method:
- `auto& context = GetEmbedderContext(EmbedderTestContextType::kVulkanContext)` becomes `auto& context = GetEmbedderContext(GetParam());`
- `builder.SetRenderTargetType(EmbedderTestBackingStoreProducer::RenderTargetType::kVulkanImage);` becomes `builder.SetRenderTargetType(GetRenderTargetFromBackend(backend, false));`
- etc

**Review by bdero — APPROVED** [review]
LGTM

File: shell/platform/embedder/tests/embedder_unittests.cc (hunk 1 above)
Comment (gaaclarke) [reply shell/platform/embedder/tests/embedder_unittests.cc:1014]: The problem is a race condition which I can't force to happen.  So this doesn't assert that the race condition is gone, but it does exercise the code to make sure that it functions.  It's possible that the failure would happen in this test infrequently without the change.

Running `embedder_unittests` this is the only way I could get this code to be hit.  The `BUILD.gn` file only compiles `embedder_gl_unittests.cc` if `test_enable_gl == true` which isn't possible to enable on macos.  Let me see if I can detangle some of this.

Comment (gaaclarke) [reply shell/platform/embedder/tests/embedder_unittests.cc]: I pulled the new tests into a `embedder_vk_unittests.cc` to mirror the ones in `embedder_gl_unittest.cc` and `embedder_metal_unittests.cc`.  It has a comment in there that says some of the tests are in `embedder_gl_unittests.cc` and I added an issue to clean them up more later (noted in comment).

**Comment by gaaclarke** [comment]
I'll try to rebase this to see if that flushes out the failure.

**Comment by gaaclarke** [comment]
I'm going to revert this.  It works fine in tests, but it's not going to work for real usage of the embedder since the user has free access to the vkQueue and thus has no way to ensure that the vkQueue isn't used concurrently.
